// convex-test/dist/index.js
/// <reference types="vite/client" />
import { getFunctionAddress } from "convex/server";
import { actionGeneric, httpActionGeneric, makeFunctionReference, mutationGeneric, queryGeneric, } from "convex/server";
import { convexToJson, jsonToConvex, } from "convex/values";
import { createHash } from "crypto";
import { compareValues } from "./compare.js";
const ROOT_COMPONENT_PATH = "";
class DatabaseFake {
    _componentPath;
    _documents = {};
    _storage = {};
    _nextQueryId = 1;
    _nextDocId = 10000;
    _lastCreationTime = 0;
    _queryResults = {};
    // TODO: Make this more robust and cleaner
    jobListener = () => { };
    // Pending writes for each level of transaction nesting.
    // Whenever a child mutation begins, a new level is created for all
    // components.
    // - When a mutation performs updates, they are staged in the last
    //   level of writes for the mutation's component.
    // - When a child mutation commits, the writes are merged up one level.
    // - When the top-level mutation commits, the writes are applied to the
    //   database.
    // - When a mutation is rolled back, last level of writes is discarded.
    _writes = [];
    _schema;
    constructor(schema, componentPath) {
        this._schema =
            schema === null
                ? null
                : {
                    schemaValidation: schema.schemaValidation,
                    tables: new Map(Object.entries(schema.tables).map(([name, tableSchema]) => [
                        name,
                        tableSchema.export(),
                    ])),
                };
        this._componentPath = componentPath;
        this.validateSchema();
    }
    // Called by TransactionManager.begin.
    startTransaction() {
        this._writes.push({});
    }
    get(id) {
        if (typeof id !== "string") {
            throw new Error(`Invalid argument \`id\` for \`db.get\`, expected string but got '${typeof id}': ${id}`);
        }
        for (let i = this._writes.length - 1; i >= 0; i--) {
            const write = this._writes[i][id];
            if (write !== undefined) {
                return write;
            }
        }
        return this._documents[id] ?? null;
    }
    // Note that this is not the format the real backend
    // uses for IDs.
    _generateId(table) {
        const id = this._nextDocId.toString() + ";" + table;
        this._nextDocId += 1;
        return id;
    }
    storeFile(storageId, blob) {
        this._storage[storageId] = blob;
    }
    getFile(storageId) {
        if (this.get(storageId) === null) {
            return null;
        }
        return this._storage[storageId];
    }
    _addWrite(id, newValue) {
        if (this._writes.length === 0) {
            throw new Error(`Write outside of transaction ${id}`);
        }
        this._writes[this._writes.length - 1][id] = newValue;
    }
    insert(table, value) {
        this._validate(table, value);
        const _id = this._generateId(table);
        const now = Date.now();
        const _creationTime = now <= this._lastCreationTime ? this._lastCreationTime + 0.001 : now;
        this._lastCreationTime = _creationTime;
        this._addWrite(_id, { ...value, _id, _creationTime });
        return _id;
    }
    patch(id, value) {
        const document = this.get(id);
        if (document === null) {
            throw new Error(`Patch on non-existent document with ID "${id}"`);
        }
        const { _id, _creationTime, ...fields } = document;
        if (value._id !== undefined && value._id !== _id) {
            throw new Error(`Provided \`_id\` field value "${value._id}" ` +
                `does not match the document ID "${_id}"`);
        }
        if (value._creationTime !== undefined &&
            value._creationTime !== _creationTime) {
            throw new Error(`Provided \`_creationTime\` field value ${value._creationTime} ` +
                `does not match the document's creation time ${_creationTime}`);
        }
        delete value["_id"];
        delete value["_creationTime"];
        const convexValue = {};
        for (const [key, v] of Object.entries(value)) {
            convexValue[key] = evaluateValue(v);
        }
        const merged = { ...fields, ...convexValue };
        this._validate(tableNameFromId(_id), merged);
        this._addWrite(id, { _id, _creationTime, ...merged });
    }
    replace(id, value) {
        const document = this.get(id);
        if (document === null) {
            throw new Error(`Replace on non-existent document with ID "${id}"`);
        }
        if (value._id !== undefined && value._id !== document._id) {
            throw new Error(`Provided \`_id\` field value "${value._id}" ` +
                `does not match the document ID "${document._id}"`);
        }
        if (value._creationTime !== undefined &&
            value._creationTime !== document._creationTime) {
            throw new Error(`Provided \`_creationTime\` field value ${value._creationTime} ` +
                `does not match the document's creation time ${document._creationTime}`);
        }
        delete value["_id"];
        delete value["_creationTime"];
        const convexValue = {};
        for (const [key, v] of Object.entries(value)) {
            convexValue[key] = evaluateValue(v);
        }
        this._validate(tableNameFromId(document._id), convexValue);
        this._addWrite(id, {
            ...convexValue,
            _id: document._id,
            _creationTime: document._creationTime,
        });
    }
    delete(id) {
        const document = this.get(id);
        if (document === null) {
            throw new Error("Delete on non-existent doc");
        }
        this._addWrite(id, null);
    }
    commit() {
        const lastWrites = this._writes.pop();
        if (!lastWrites) {
            throw new Error("Transaction already committed or rolled back");
        }
        for (const [id, write] of Object.entries(lastWrites)) {
            const _id = id;
            if (this._writes.length === 0) {
                if (write === null) {
                    delete this._documents[_id];
                }
                else {
                    this._documents[_id] = write;
                }
            }
            else {
                this._addWrite(_id, write);
            }
        }
    }
    rollbackWrites() {
        if (this._writes.length === 0) {
            throw new Error("Transaction already committed or rolled back");
        }
        this._writes.pop();
    }
    _validate(tableName, doc) {
        if (this._schema === null || !this._schema.schemaValidation) {
            return;
        }
        const validator = this._schema.tables.get(tableName)?.documentType;
        if (validator === undefined) {
            return;
        }
        validateValidator(validator, doc);
    }
    startQuery(query) {
        const id = this._nextQueryId;
        const results = this._evaluateQuery(query);
        this._queryResults[id] = results;
        this._nextQueryId += 1;
        return id;
    }
    queryNext(queryId) {
        const results = this._queryResults[queryId];
        if (results === undefined) {
            throw new Error("Bad queryId");
        }
        if (results.length === 0) {
            return { value: null, done: true };
        }
        else {
            return { value: results.shift(), done: false };
        }
    }
    paginate({ query, cursor, pageSize, }) {
        const queryId = this.startQuery(query);
        const page = [];
        let isInPage = cursor === null;
        let isDone = false;
        let continueCursor = null;
        for (;;) {
            const { value, done } = this.queryNext(queryId);
            if (done) {
                isDone = true;
                // We have reached the end of the query. Return a cursor that indicates
                // "end query", which we can do with any string that isn't a valid _id.
                continueCursor = "_end_cursor";
                break;
            }
            if (isInPage) {
                page.push(value);
                if (page.length >= pageSize) {
                    continueCursor = value._id;
                    break;
                }
            }
            if (value._id === cursor) {
                isInPage = true;
            }
        }
        return {
            page,
            isDone,
            continueCursor,
        };
    }
    _iterateDocs(tableName, callback) {
        const isInTable = (id) => tableNameFromId(id) === tableName;
        const ids = new Set(Object.keys(this._documents).filter(isInTable));
        for (let i = 0; i < this._writes.length; i++) {
            for (const id of Object.keys(this._writes[i]).filter(isInTable)) {
                ids.add(id);
            }
        }
        for (const id of ids) {
            const document = this.get(id);
            if (document !== null) {
                callback(document);
            }
        }
    }
    _evaluateQuery(query) {
        const source = query.source;
        let results = [];
        let fieldPathsToSortBy;
        let order;
        switch (source.type) {
            case "FullTableScan": {
                const tableName = source.tableName;
                this._iterateDocs(tableName, (doc) => {
                    results.push(doc);
                });
                order = source.order ?? "asc";
                fieldPathsToSortBy = ["_creationTime"];
                break;
            }
            case "IndexRange": {
                const [tableName, indexName] = source.indexName.split(".");
                order = source.order ?? "asc";
                let fields;
                if (indexName === "by_creation_time") {
                    fields = ["_creationTime", "_id"];
                }
                else if (indexName === "by_id") {
                    fields = ["_id"];
                }
                else {
                    const indexes = this._schema?.tables.get(tableName)?.indexes;
                    const index = indexes?.find(({ indexDescriptor }) => indexDescriptor === indexName);
                    if (index === undefined) {
                        throw new Error(`Cannot use index "${indexName}" for table "${tableName}" because ` +
                            `it is not declared in the schema.`);
                    }
                    fields = index.fields.concat(["_creationTime", "_id"]);
                }
                fieldPathsToSortBy = fields;
                validateIndexRangeExpression(source, fields);
                this._iterateDocs(tableName, (doc) => {
                    if (source.range.every((filter) => evaluateRangeFilter(doc, filter))) {
                        results.push(doc);
                    }
                });
                break;
            }
            case "Search": {
                const [tableName] = source.indexName.split(".");
                this._iterateDocs(tableName, (doc) => {
                    if (source.filters.every((filter) => evaluateSearchFilter(doc, filter))) {
                        results.push(doc);
                    }
                });
                fieldPathsToSortBy = [];
                order = "asc";
                break;
            }
        }
        const filters = query.operators
            .filter((operator) => "filter" in operator)
            .map((operator) => operator.filter);
        const limit = query.operators.filter((operator) => "limit" in operator)[0] ?? null;
        results = results.filter((v) => filters.every((f) => evaluateFilter(v, f)));
        results.sort((a, b) => {
            const orderMultiplier = order === "asc" ? 1 : -1;
            let v = 0;
            for (const fp of fieldPathsToSortBy) {
                v = compareValues(evaluateFieldPath(fp, a), evaluateFieldPath(fp, b));
                if (v !== 0) {
                    return v * orderMultiplier;
                }
            }
            return v * orderMultiplier;
        });
        if (limit !== null) {
            return results.slice(0, limit.limit);
        }
        return results;
    }
    jobFinished(jobId) {
        this.jobListener(jobId);
    }
    vectorSearch(tableAndIndexName, vector, expressions, limit) {
        const results = [];
        const [tableName, indexName] = tableAndIndexName.split(".");
        this._iterateDocs(tableName, (doc) => {
            if (expressions.every((filter) => evaluateFilter(doc, filter))) {
                results.push(doc);
            }
        });
        const vectorIndexes = this._schema?.tables.get(tableName)?.vectorIndexes;
        const vectorIndex = vectorIndexes?.find(({ indexDescriptor }) => indexDescriptor === indexName);
        if (vectorIndex === undefined) {
            throw new Error(`Cannot use vector index "${indexName}" for table "${tableName}" because ` +
                `it is not declared in the schema.`);
        }
        const { vectorField } = vectorIndex;
        const idsAndScores = results.map((doc) => {
            const score = cosineSimilarity(vector, doc[vectorField]);
            return { _id: doc._id, _score: score };
        });
        idsAndScores.sort((a, b) => b._score - a._score);
        return idsAndScores.slice(0, limit);
    }
    validateSchema() {
        this._schema?.tables.forEach((table, tableName) => {
            if (!isValidIdentifier(tableName)) {
                throw new Error(`Table names must be valid identifiers, got "${tableName}"`);
            }
            validateFieldNames(table.documentType);
            table.indexes.forEach(({ indexDescriptor }) => {
                if (!isValidIdentifier(indexDescriptor)) {
                    throw new Error(`Index names must be valid identifiers, got "${indexDescriptor}"`);
                }
            });
        });
    }
    get componentPath() {
        return this._componentPath;
    }
}
function tableNameFromId(id) {
    const parts = id.split(";");
    if (parts.length !== 2) {
        return null;
    }
    return id.split(";")[1];
}
function isSimpleObject(value) {
    const isObject = typeof value === "object";
    const prototype = Object.getPrototypeOf(value);
    const isSimple = prototype === null ||
        prototype === Object.prototype ||
        // Objects generated from other contexts (e.g. across Node.js `vm` modules) will not satisfy the previous
        // conditions but are still simple objects.
        prototype?.constructor?.name === "Object";
    return isObject && isSimple;
}
function evaluateFieldPath(fieldPath, document) {
    const pathParts = fieldPath.split(".");
    let result = document;
    for (const p of pathParts) {
        result =
            result !== undefined && result !== null && isSimpleObject(result)
                ? result[p]
                : undefined;
    }
    return result;
}
function evaluateFilter(document, filter) {
    if (filter.$eq !== undefined) {
        return compareValues(evaluateFilter(document, filter.$eq[0]), evaluateFilter(document, filter.$eq[1])) === 0;
    }
    if (filter.$neq !== undefined) {
        return compareValues(evaluateFilter(document, filter.$neq[0]), evaluateFilter(document, filter.$neq[1])) !== 0;
    }
    if (filter.$and !== undefined) {
        return filter.$and.every((child) => evaluateFilter(document, child));
    }
    if (filter.$or !== undefined) {
        return filter.$or.some((child) => evaluateFilter(document, child));
    }
    if (filter.$not !== undefined) {
        return !evaluateFilter(document, filter.$not);
    }
    if (filter.$gt !== undefined) {
        return (evaluateFilter(document, filter.$gt[0]) >
            evaluateFilter(document, filter.$gt[1]));
    }
    if (filter.$gte !== undefined) {
        return (evaluateFilter(document, filter.$gte[0]) >=
            evaluateFilter(document, filter.$gte[1]));
    }
    if (filter.$lt !== undefined) {
        return (evaluateFilter(document, filter.$lt[0]) <
            evaluateFilter(document, filter.$lt[1]));
    }
    if (filter.$lte !== undefined) {
        return (evaluateFilter(document, filter.$lte[0]) <=
            evaluateFilter(document, filter.$lte[1]));
    }
    if (filter.$add !== undefined) {
        return (evaluateFilter(document, filter.$add[0]) +
            evaluateFilter(document, filter.$add[1]));
    }
    if (filter.$sub !== undefined) {
        return (evaluateFilter(document, filter.$sub[0]) -
            evaluateFilter(document, filter.$sub[1]));
    }
    if (filter.$mul !== undefined) {
        return (evaluateFilter(document, filter.$mul[0]) *
            evaluateFilter(document, filter.$mul[1]));
    }
    if (filter.$div !== undefined) {
        return (evaluateFilter(document, filter.$div[0]) /
            evaluateFilter(document, filter.$div[1]));
    }
    if (filter.$mod !== undefined) {
        return (evaluateFilter(document, filter.$mod[0]) %
            evaluateFilter(document, filter.$mod[1]));
    }
    if (filter.$field !== undefined) {
        return evaluateFieldPath(filter.$field, document);
    }
    if (filter.$literal !== undefined) {
        return evaluateValue(filter.$literal);
    }
    throw new Error(`not implemented: ${JSON.stringify(filter)}`);
}
function evaluateRangeFilter(document, expr) {
    const result = evaluateFieldPath(expr.fieldPath, document);
    const value = evaluateValue(expr.value);
    switch (expr.type) {
        case "Eq":
            return compareValues(result, value) === 0;
        case "Gt":
            return compareValues(result, value) > 0;
        case "Gte":
            return compareValues(result, value) >= 0;
        case "Lt":
            return compareValues(result, value) < 0;
        case "Lte":
            return compareValues(result, value) <= 0;
    }
}
function evaluateValue(value) {
    if (typeof value === "object" && value !== null && "$undefined" in value) {
        return undefined;
    }
    return jsonToConvex(value);
}
function evaluateSearchFilter(document, filter) {
    const result = evaluateFieldPath(filter.fieldPath, document);
    switch (filter.type) {
        case "Eq":
            return compareValues(result, filter.value) === 0;
        case "Search":
            return result
                .split(/\s/)
                .some((word) => word.startsWith(filter.value));
    }
}
function cosineSimilarity(vecA, vecB) {
    let dotProduct = 0.0;
    let normA = 0.0;
    let normB = 0.0;
    for (let i = 0; i < vecA.length; i++) {
        dotProduct += vecA[i] * vecB[i];
        normA += vecA[i] * vecA[i];
        normB += vecB[i] * vecB[i];
    }
    if (normA === 0 || normB === 0) {
        return 0;
    }
    else {
        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }
}
// An index range expression is always a chained list of:
//
// 0 or more equality expressions defined with .eq.
// [Optionally] A lower bound expression defined with .gt or .gte.
// [Optionally] An upper bound expression defined with .lt or .lte.
function validateIndexRangeExpression(source, fields) {
    let fieldIndex = 0;
    let state = "eq";
    for (const [filterIndex, filter] of source.range.entries()) {
        if (state === "done") {
            throw new Error(`Incorrect operator used in \`withIndex\`, cannot chain ` +
                `more operators after both \`.gt\` and \`.lt\` were already used, ` +
                `got \`${printIndexOperator(filter)}\`.`);
        }
        const filterType = filter.type == "Gt" || filter.type == "Gte"
            ? "gt"
            : filter.type == "Lt" || filter.type == "Lte"
                ? "lt"
                : "eq";
        switch (`${state}|${filterType}`) {
            // Allow to operate on the current indexed field
            case "eq|eq":
            case "eq|gt":
            case "eq|lt":
                if (filter.fieldPath === fields[fieldIndex]) {
                    fieldIndex += 1;
                    state = filterType;
                    continue;
                }
                throw new Error(`Incorrect field used in \`withIndex\`, ` +
                    `expected "${fields[fieldIndex]}", got "${filter.fieldPath}"`);
            // Allow to operate on the previous field (gt and lt must operate on same field)
            case "lt|gt":
            case "gt|lt":
                if (fieldIndex > 0 && filter.fieldPath === fields[fieldIndex - 1]) {
                    state = "done";
                    continue;
                }
                throw new Error(`Incorrect field used in \`withIndex\`, ` +
                    `\`.gt\` and \`.lt\` must operate on the same field, ` +
                    `expected "${fields[fieldIndex - 1]}", got "${filter.fieldPath}"`);
            default:
                throw new Error(`Incorrect operator used in \`withIndex\`, ` +
                    `cannot chain \`.${filter.type.toLowerCase()}()\` ` +
                    `after \`.${source.range[filterIndex - 1].type.toLowerCase()}()\``);
        }
    }
}
function printIndexOperator(filter) {
    return `.${filter.type.toLowerCase()}(${filter.fieldPath}, ${JSON.stringify(filter.value)})`;
}
function validateFieldNames(validator) {
    validator.type === "object" &&
        Object.keys(validator.value).forEach((fieldName) => {
            if (!isValidIdentifier(fieldName)) {
                throw new Error(`Field names must be valid identifiers, got "${fieldName}"`);
            }
        });
    validator.type === "union" && validator.value.forEach(validateFieldNames);
}
function isValidIdentifier(name) {
    return /^[a-zA-Z][a-zA-Z0-9_]*$/.test(name);
}
function validateValidator(validator, value) {
    switch (validator.type) {
        case "null": {
            if (value !== null) {
                throw new Error(`Validator error: Expected \`null\`, got \`${value}\``);
            }
            return;
        }
        case "number": {
            if (typeof value !== "number") {
                throw new Error(`Validator error: Expected \`number\`, got \`${value}\``);
            }
            return;
        }
        case "bigint": {
            if (typeof value !== "bigint") {
                throw new Error(`Validator error: Expected \`bigint\`, got \`${value}\``);
            }
            return;
        }
        case "boolean": {
            if (typeof value !== "boolean") {
                throw new Error(`Validator error: Expected \`boolean\`, got \`${value}\``);
            }
            return;
        }
        case "string": {
            if (typeof value !== "string") {
                throw new Error(`Validator error: Expected \`string\`, got \`${value}\``);
            }
            return;
        }
        case "bytes": {
            if (!(value instanceof ArrayBuffer)) {
                throw new Error(`Validator error: Expected \`ArrayBuffer\`, got \`${value}\``);
            }
            return;
        }
        case "any": {
            return;
        }
        case "literal": {
            if (value !== validator.value) {
                throw new Error(`Validator error: Expected \`${validator.value}\`, got \`${value}\``);
            }
            return;
        }
        case "id": {
            if (typeof value !== "string") {
                throw new Error(`Validator error: Expected \`string\`, got \`${value}\``);
            }
            if (tableNameFromId(value) !== validator.tableName) {
                throw new Error(`Validator error: Expected ID for table "${validator.tableName}", got \`${value}\``);
            }
            return;
        }
        case "array": {
            if (!Array.isArray(value)) {
                throw new Error(`Validator error: Expected \`Array\`, got \`${value}\``);
            }
            for (const v of value) {
                validateValidator(validator.value, v);
            }
            return;
        }
        case "object": {
            if (typeof value !== "object") {
                throw new Error(`Validator error: Expected \`object\`, got \`${value}\``);
            }
            if (!isSimpleObject(value)) {
                throw new Error(`Validator error: Expected a plain old JavaScript \`object\`, got \`${value}\``);
            }
            for (const [k, { fieldType, optional }] of Object.entries(validator.value)) {
                if (value[k] === undefined) {
                    if (!optional) {
                        throw new Error(`Validator error: Missing required field \`${k}\` in object`);
                    }
                }
                else {
                    validateValidator(fieldType, value[k]);
                }
            }
            for (const k of Object.keys(value)) {
                if (validator.value[k] === undefined) {
                    throw new Error(`Validator error: Unexpected field \`${k}\` in object`);
                }
            }
            return;
        }
    }
}
function syscallImpl() {
    return (op, jsonArgs) => {
        const args = JSON.parse(jsonArgs);
        const db = getDb();
        switch (op) {
            case "1.0/queryStream": {
                const { query } = args;
                const queryId = db.startQuery(query);
                return JSON.stringify({ queryId });
            }
            case "1.0/queryCleanup": {
                return JSON.stringify({});
            }
            case "1.0/db/normalizeId": {
                const idString = args.idString;
                const isInTable = idString.endsWith(`;${args.table}`);
                return JSON.stringify({
                    id: isInTable ? idString : null,
                });
            }
            default: {
                throw new Error(`\`convexTest\` does not support syscall: "${op}"`);
            }
        }
    };
}
class AuthFake {
    _userIdentity;
    constructor(_userIdentity = null) {
        this._userIdentity = _userIdentity;
    }
    getUserIdentity() {
        return Promise.resolve(this._userIdentity);
    }
}
function asyncSyscallImpl() {
    return async (op, jsonArgs) => {
        const args = JSON.parse(jsonArgs);
        const db = getDb();
        switch (op) {
            case "1.0/get": {
                const doc = db.get(args.id);
                return JSON.stringify(convexToJson(doc));
            }
            case "1.0/queryStreamNext": {
                const { value, done } = db.queryNext(args.queryId);
                return JSON.stringify(convexToJson({ value, done }));
            }
            case "1.0/queryPage": {
                const { query, cursor, pageSize } = args;
                const { page, isDone, continueCursor } = db.paginate({
                    query,
                    cursor,
                    pageSize,
                });
                return JSON.stringify(convexToJson({ page, isDone, continueCursor }));
            }
            case "1.0/insert": {
                const _id = db.insert(args.table, jsonToConvex(args.value));
                return JSON.stringify({ _id });
            }
            case "1.0/shallowMerge": {
                const { id, value } = args;
                db.patch(id, value);
                return JSON.stringify({});
            }
            case "1.0/replace": {
                const { id, value } = args;
                db.replace(id, value);
                return JSON.stringify({});
            }
            case "1.0/remove": {
                const { id } = args;
                db.delete(id);
                return JSON.stringify({});
            }
            case "1.0/actions/query": {
                const { name, args: queryArgs } = args;
                return JSON.stringify(convexToJson(await withAuth().query(makeFunctionReference(name), queryArgs)));
            }
            case "1.0/actions/mutation": {
                const { name, args: mutationArgs } = args;
                return JSON.stringify(convexToJson(await withAuth().mutation(makeFunctionReference(name), mutationArgs)));
            }
            case "1.0/actions/action": {
                const { name, args: actionArgs } = args;
                return JSON.stringify(convexToJson(await withAuth().action(makeFunctionReference(name), actionArgs)));
            }
            case "1.0/runUdf": {
                const { udfType, name, reference, functionHandle, args: udfArgs, } = args;
                const functionPath = await getFunctionPathFromAddress({
                    name,
                    reference,
                    functionHandle,
                });
                if (udfType === "query") {
                    return JSON.stringify(convexToJson(await withAuth().queryFromPath(functionPath, /* isNested */ true, udfArgs)));
                }
                if (udfType === "mutation") {
                    return JSON.stringify(convexToJson(await withAuth().mutationFromPath(functionPath, /* isNested */ true, udfArgs)));
                }
                throw new Error(`\`convexTest\` does not support udf type: "${udfType}"`);
            }
            case "1.0/createFunctionHandle": {
                const { name, reference, functionHandle, } = args;
                const functionPath = await getFunctionPathFromAddress({
                    name,
                    reference,
                    functionHandle,
                });
                const handle = createFunctionHandle(functionPath);
                return JSON.stringify(handle);
            }
            case "1.0/actions/schedule": {
                return await withAuth().run(async () => {
                    return await getSyscalls().asyncSyscall("1.0/schedule", jsonArgs);
                });
            }
            case "1.0/schedule": {
                const { name, reference, functionHandle, args: fnArgs, ts: tsInSecs, } = args;
                const functionPath = await getFunctionPathFromAddress({
                    name,
                    reference,
                    functionHandle,
                });
                const jobId = db.insert("_scheduled_functions", {
                    args: [fnArgs],
                    name: functionPath.udfPath,
                    scheduledTime: tsInSecs * 1000,
                    state: { kind: "pending" },
                });
                const componentPath = getCurrentComponentPath();
                setTimeout((async () => {
                    const canceled = await withAuth().runInComponent(componentPath, async () => {
                        const job = db.get(jobId);
                        if (job.state.kind === "canceled") {
                            return true;
                        }
                        if (job.state.kind !== "pending") {
                            throw new Error(`\`convexTest\` invariant error: Unexpected scheduled function state when starting it: ${job.state.kind}`);
                        }
                        db.patch(jobId, { state: { kind: "inProgress" } });
                        return false;
                    });
                    if (canceled) {
                        return;
                    }
                    try {
                        await withAuth().fun(functionPath, fnArgs);
                    }
                    catch (error) {
                        console.error(`Error when running scheduled function ${name}`, error);
                        await withAuth().runInComponent(componentPath, async () => {
                            db.patch(jobId, {
                                state: { kind: "failed" },
                                completedTime: Date.now(),
                            });
                        });
                        db.jobFinished(jobId);
                        return;
                    }
                    await withAuth().runInComponent(componentPath, async () => {
                        const job = db.get(jobId);
                        if (job.state.kind !== "inProgress") {
                            throw new Error(`\`convexTest\` invariant error: Unexpected scheduled function state after it finished running: ${job.state.kind}`);
                        }
                        db.patch(jobId, { state: { kind: "success" } });
                    });
                    db.jobFinished(jobId);
                }), tsInSecs * 1000 - Date.now());
                return JSON.stringify(convexToJson(jobId));
            }
            case "1.0/actions/cancel_job": {
                await withAuth().run(async () => {
                    await getSyscalls().asyncSyscall("1.0/cancel_job", jsonArgs);
                });
                return JSON.stringify({});
            }
            case "1.0/actions/vectorSearch": {
                const { query: { indexName, limit, vector, expressions }, } = args;
                const results = db.vectorSearch(indexName, vector, 
                // Probably an unintentional implementation in Convex
                // where expressions is only a single expression
                expressions === null ? [] : [expressions], limit);
                return JSON.stringify(convexToJson({ results }));
            }
            case "1.0/cancel_job": {
                const { id } = args;
                db.patch(id, { state: { kind: "canceled" } });
                return JSON.stringify({});
            }
            case "1.0/storageDelete": {
                const { storageId } = args;
                await writeToDatabase(async (db) => {
                    db.delete(storageId);
                });
                return JSON.stringify({});
            }
            case "1.0/storageGetUrl": {
                const { storageId } = args;
                const metadata = db.get(storageId);
                if (metadata === null) {
                    return JSON.stringify(null);
                }
                const { sha256 } = metadata;
                // In the real backend the URL ofc isn't the sha
                const url = "https://some-deployment.convex.cloud/api/storage/" +
                    sha256;
                return JSON.stringify(convexToJson(url));
            }
            case "1.0/storageGenerateUploadUrl": {
                // In the real backend the token is cryptographically secure
                const url = "https://some-deployment.convex.cloud/api/storage/upload?token=" +
                    Math.random();
                return JSON.stringify(convexToJson(url));
            }
            default: {
                throw new Error(`\`convexTest\` does not support async syscall: "${op}"`);
            }
        }
    };
}
function jsSyscallImpl() {
    return async (op, args) => {
        const db = getDb();
        switch (op) {
            case "storage/storeBlob": {
                const { blob } = args;
                const storageId = await writeToDatabase(async (db) => {
                    return db.insert("_storage", {
                        size: blob.size,
                        sha256: await blobSha(blob),
                    });
                });
                db.storeFile(storageId, blob);
                return storageId;
            }
            case "storage/getBlob": {
                const { storageId } = args;
                return db.getFile(storageId);
            }
            default: {
                throw new Error(`\`convexTest\` does not support js syscall: "${op}"`);
            }
        }
    };
}
// If we're in action, wrap the write in a transaction.
async function writeToDatabase(impl) {
    const db = getDb();
    if (!getTransactionManager().isInTransaction()) {
        return await withAuth().run(async () => {
            return await impl(db);
        });
    }
    else {
        return await impl(db);
    }
}
async function blobSha(blob) {
    const arrayBuffer = await blob.arrayBuffer();
    const sha256 = createHash("sha256");
    sha256.update(Buffer.from(arrayBuffer));
    return sha256.digest("base64");
}
async function waitForInProgressScheduledFunctions() {
    let hadScheduledFunctions = false;
    for (const componentPath of Object.keys(getConvexGlobal().components)) {
        const inProgressJobs = (await withAuth().runInComponent(componentPath, async (ctx) => {
            return (await ctx.db.system.query("_scheduled_functions").collect()).filter((job) => job.state.kind === "inProgress");
        }));
        let numRemaining = inProgressJobs.length;
        if (numRemaining === 0) {
            continue;
        }
        hadScheduledFunctions = true;
        await new Promise((resolve) => {
            getDbForComponent(componentPath).jobListener = () => {
                numRemaining -= 1;
                if (numRemaining === 0) {
                    resolve();
                }
            };
        });
    }
    return hadScheduledFunctions;
}
function getComponentInfo(componentPath) {
    const convex = getConvexGlobal();
    if (convex.components[componentPath] === undefined) {
        throw new Error(`Component "${componentPath}" is not registered. Call "t.registerComponent".`);
    }
    return convex.components[componentPath];
}
function getDbForComponent(componentPath) {
    return getComponentInfo(componentPath).db;
}
function getDb() {
    return getDbForComponent(getCurrentComponentPath());
}
function getTransactionManager() {
    return getConvexGlobal().transactionManager;
}
function getCurrentComponentPath() {
    const functionStack = getTransactionManager().functionStack;
    const currentFunctionPath = functionStack[functionStack.length - 1];
    return currentFunctionPath?.componentPath ?? ROOT_COMPONENT_PATH;
}
function getModules() {
    return getModulesForComponent(getCurrentComponentPath());
}
function getModulesForComponent(componentPath) {
    return getComponentInfo(componentPath).modules;
}
function getSyscalls() {
    return global.Convex;
}
function moduleCache(specifiedModules) {
    const modules = specifiedModules ?? import.meta.glob("../../../convex/**/*.*s");
    const prefix = findModulesRoot(Object.keys(modules), specifiedModules !== undefined);
    const modulesWithoutExtension = Object.fromEntries(Object.entries(modules).map(([path, module]) => [
        path.replace(/\.[^.]+$/, ""),
        module,
    ]));
    return async (path) => {
        const module = modulesWithoutExtension[prefix + path];
        if (module === undefined) {
            throw new Error(`Could not find module for: "${path}"`);
        }
        return await module();
    };
}
function findModulesRoot(modulesPaths, userProvidedModules) {
    const generatedFilePath = modulesPaths.find((path) => path.includes("_generated"));
    if (generatedFilePath !== undefined) {
        return generatedFilePath.split("_generated", 2)[0];
    }
    throw new Error('Could not find the "_generated" directory, make sure to run ' +
        "`npx convex dev` or `npx convex codegen`. " +
        (userProvidedModules
            ? "Make sure your `import.meta.glob` includes the files in the " +
                '"_generated" directory'
            : "If your Convex functions aren't defined in a directory " +
                'called "convex" sibling to your node_modules, ' +
                "provide the second argument to `convexTest`"));
}
function getConvexGlobal() {
    return global.Convex;
}
function setConvexGlobal(convex) {
    if (getConvexGlobal()) {
        if (getTransactionManager().isInTransaction()) {
            throw new Error("test began while previous transaction was still open");
        }
    }
    global.Convex = convex;
}
class TransactionManager {
    // The DatabaseFake is used in the Convex global,
    // and so it restricts `convexTest` to run one function
    // at a time.
    // We force sequential execution to make sure actions
    // can run mutations in parallel.
    _waitOnCurrentFunction = null;
    _markTransactionDone = null;
    functionStack = [];
    async begin(functionPath, isNested) {
        // Take a lock only for the top-level of each transaction.
        // Nested transactions are not isolated so if you `Promise.all` on multiple
        // `ctx.runMutation` or `ctx.runQuery` calls, they won't be serialized.
        if (!isNested) {
            // Note the loop is important, as the current promise might resolve and a
            // new transaction could start before we get woken up.
            // This is the standard pattern for condition variables:
            // https://en.wikipedia.org/wiki/Monitor_(synchronization)
            while (this._waitOnCurrentFunction !== null) {
                await this._waitOnCurrentFunction;
            }
            this._waitOnCurrentFunction = new Promise((resolve) => {
                this._markTransactionDone = resolve;
            });
        }
        this.functionStack.push(functionPath);
        const convex = getConvexGlobal();
        for (const component of Object.values(convex.components)) {
            component.db.startTransaction();
        }
    }
    beginAction(functionPath) {
        this.functionStack.push(functionPath);
    }
    finishAction() {
        this.functionStack.pop();
    }
    // Used to distinguish between mutation and action execution
    // environment.
    isInTransaction() {
        return this._waitOnCurrentFunction !== null;
    }
    commit(isNested) {
        const convex = getConvexGlobal();
        for (const component of Object.values(convex.components)) {
            component.db.commit();
        }
        this._endTransaction(isNested);
    }
    rollback(isNested) {
        const convex = getConvexGlobal();
        for (const component of Object.values(convex.components)) {
            component.db.rollbackWrites();
        }
        this._endTransaction(isNested);
    }
    _endTransaction(isNested) {
        if (this._markTransactionDone === null) {
            throw new Error("Transaction not started");
        }
        this.functionStack.pop();
        if (!isNested) {
            this._waitOnCurrentFunction = null;
            this._markTransactionDone();
            this._markTransactionDone = null;
        }
    }
}
/**
 * Call this function at the start of each of your tests.
 *
 * @param schema The default export from your "schema.ts" file.
 * @param modules If you have a custom `functions` path
 *   in convex.json, provide the module map with your functions
 *   by calling `import.meta.glob` with the appropriate glob pattern
 *   for paths relative to the file where you call it.
 * @returns an object which is by convention stored in the `t` variable
 *   and which provides methods for exercising your Convex functions.
 */
export const convexTest = (schema, 
// For example `import.meta.glob("./**/*.*s")`
modules) => {
    const rootDb = new DatabaseFake(schema ?? null, ROOT_COMPONENT_PATH);
    setConvexGlobal({
        components: {
            [ROOT_COMPONENT_PATH]: {
                db: rootDb,
                modules: moduleCache(modules),
            },
        },
        transactionManager: new TransactionManager(),
        syscall: syscallImpl(),
        asyncSyscall: asyncSyscallImpl(),
        jsSyscall: jsSyscallImpl(),
    });
    return {
        withIdentity(identity) {
            const subject = identity.subject ?? "" + simpleHash(JSON.stringify(identity));
            const issuer = identity.issuer ?? "https://convex.test";
            const tokenIdentifier = identity.tokenIdentifier ?? `${issuer}|${subject}`;
            return withAuth(new AuthFake({ ...identity, subject, issuer, tokenIdentifier }));
        },
        ...withAuth(),
        registerComponent(componentPath, schema, glob) {
            const componentInfo = {
                db: new DatabaseFake(schema, componentPath),
                modules: moduleCache(glob),
            };
            getConvexGlobal().components[componentPath] = componentInfo;
        },
    };
};
function withAuth(auth = new AuthFake()) {
    const runTransaction = async (handler, args, extraCtx = {}, functionPath, isNested) => {
        const m = mutationGeneric({
            handler: (ctx, a) => {
                const testCtx = { ...ctx, auth, ...extraCtx };
                return handler(testCtx, a);
            },
        });
        const transactionManager = getTransactionManager();
        await transactionManager.begin(functionPath, isNested);
        try {
            const rawResult = await m.invokeMutation(JSON.stringify(convexToJson([parseArgs(args)])));
            transactionManager.commit(isNested);
            return jsonToConvex(JSON.parse(rawResult));
        }
        catch (e) {
            transactionManager.rollback(isNested);
            throw e;
        }
    };
    const byTypeWithPath = {
        queryFromPath: async (functionPath, isNested, args) => {
            const func = await getFunctionFromPath(functionPath, "query");
            validateValidator(JSON.parse(func.exportArgs()), args ?? {});
            const q = queryGeneric({
                handler: (ctx, a) => {
                    const testCtx = { ...ctx, auth };
                    return func(testCtx, a);
                },
            });
            const transactionManager = getTransactionManager();
            await transactionManager.begin(functionPath, isNested);
            try {
                const rawResult = await q.invokeQuery(JSON.stringify(convexToJson([parseArgs(args)])));
                return jsonToConvex(JSON.parse(rawResult));
            }
            finally {
                transactionManager.rollback(isNested);
            }
        },
        mutationFromPath: async (functionPath, isNested, args) => {
            const func = await getFunctionFromPath(functionPath, "mutation");
            validateValidator(JSON.parse(func.exportArgs()), args ?? {});
            return await runTransaction(func, args, {}, functionPath, isNested);
        },
        actionFromPath: async (functionPath, args) => {
            const func = await getFunctionFromPath(functionPath, "action");
            validateValidator(JSON.parse(func.exportArgs()), args ?? {});
            const a = actionGeneric({
                handler: (ctx, a) => {
                    const testCtx = {
                        ...ctx,
                        runQuery: byType.query,
                        runMutation: byType.mutation,
                        runAction: byType.action,
                        auth,
                    };
                    return func(testCtx, a);
                },
            });
            getTransactionManager().beginAction(functionPath);
            // Real backend uses different ID format
            const requestId = "" + Math.random();
            const rawResult = await a.invokeAction(requestId, JSON.stringify(convexToJson([parseArgs(args)])));
            getTransactionManager().finishAction();
            return jsonToConvex(JSON.parse(rawResult));
        },
    };
    const byType = {
        query: async (functionReference, args) => {
            const functionPath = await getFunctionPathFromReference(functionReference);
            return await byTypeWithPath.queryFromPath(functionPath, /* isNested */ false, args);
        },
        mutation: async (functionReference, args) => {
            const functionPath = await getFunctionPathFromReference(functionReference);
            return await byTypeWithPath.mutationFromPath(functionPath, /* isNested */ false, args);
        },
        action: async (functionReference, args) => {
            const functionPath = await getFunctionPathFromReference(functionReference);
            return await byTypeWithPath.actionFromPath(functionPath, args);
        },
    };
    const run = async (componentPath, handler) => {
        // Grab StorageActionWriter from action ctx
        const a = actionGeneric({
            handler: async ({ storage }) => {
                return await runTransaction(handler, {}, { storage }, 
                // Fake mutation path.
                {
                    componentPath,
                    udfPath: "custom",
                }, false);
            },
        });
        // Real backend uses different ID format
        const requestId = "" + Math.random();
        const rawResult = await a.invokeAction(requestId, JSON.stringify(convexToJson([{}])));
        return jsonToConvex(JSON.parse(rawResult));
    };
    return {
        ...byType,
        ...byTypeWithPath,
        runInComponent: run,
        run: async (handler) => {
            return await run(getCurrentComponentPath(), handler);
        },
        fun: async (functionPath, args) => {
            const func = await getFunctionFromPath(functionPath, "any");
            if (func.isQuery) {
                return await byTypeWithPath.queryFromPath(functionPath, /* isNested */ false, args);
            }
            if (func.isMutation) {
                return await byTypeWithPath.mutationFromPath(functionPath, /* isNested */ false, args);
            }
            if (func.isAction) {
                return await byTypeWithPath.actionFromPath(functionPath, args);
            }
        },
        fetch: async (path, init) => {
            const router = (await getModules()("http"))["default"];
            if (!path.startsWith("/")) {
                throw new Error(`Path given to \`t.fetch\` must start with a \`/\``);
            }
            const url = new URL(`https://some.convex.site${path}`);
            const found = router.lookup(url.pathname, init?.method ?? "GET");
            if (!found) {
                return new Response(`No HttpAction routed for ${url.pathname}`, {
                    status: 404,
                });
            }
            const [func] = found;
            const a = httpActionGeneric((ctx, a) => {
                const testCtx = {
                    ...ctx,
                    runQuery: byType.query,
                    runMutation: byType.mutation,
                    runAction: byType.action,
                    auth,
                };
                // TODO: Remove `any`, it's needed because of a bug in Convex types
                return func(testCtx, a);
            });
            const response = await a.invokeHttpAction(new Request(url, init));
            return response;
        },
        // This is needed because when we execute functions
        // we are performing dynamic `import`s, and those
        // are real work that cannot be force-awaited.
        finishInProgressScheduledFunctions: async () => {
            await waitForInProgressScheduledFunctions();
        },
        finishAllScheduledFunctions: async (advanceTimers, maxIterations = 100) => {
            // Wait for all scheduled functions to finish, advancing time in between
            // each function.
            // Stop after a fixed number of iterations to avoid infinite loops.
            for (let i = 0; i < maxIterations; i++) {
                advanceTimers();
                const hadScheduledFunctions = await waitForInProgressScheduledFunctions();
                if (!hadScheduledFunctions) {
                    return;
                }
            }
            throw new Error("finishAllScheduledFunctions: too many iterations. "
                + "Check for infinitely recursive scheduled functions, "
                + "or increase maxIterations.");
        },
    };
}
function parseArgs(args) {
    if (args === undefined) {
        return {};
    }
    if (!isSimpleObject(args)) {
        throw new Error(`The arguments to a Convex function must be an object. Received: ${args}`);
    }
    return args;
}
function createFunctionHandle(functionPath) {
    return `function://${functionPath.componentPath};${functionPath.udfPath}`;
}
function parseFunctionHandle(handle) {
    const [componentPath, udfPath] = handle.split("function://")[1].split(";");
    return { componentPath, udfPath };
}
async function getFunctionPathFromAddress(functionAddress) {
    if (functionAddress.functionHandle !== undefined) {
        return parseFunctionHandle(functionAddress.functionHandle);
    }
    if (functionAddress.name !== undefined) {
        return {
            udfPath: functionAddress.name,
            componentPath: getCurrentComponentPath(),
        };
    }
    if (functionAddress.reference !== undefined) {
        // "_reference/childComponent/aggregate/path/to/file/functionName" -> "aggregate/path/to/file"
        const childComponentName = functionAddress.reference.split("/")[2];
        let componentPath = childComponentName;
        if (getCurrentComponentPath().length > 0) {
            componentPath = `${getCurrentComponentPath()}/${componentPath}`;
        }
        // "_reference/childComponent/aggregate/path/to/file/functionName" -> "path/to/file/functionName"
        const functionNameWithSlashes = functionAddress.reference
            .split("/")
            .slice(3)
            .join("/");
        // "path/to/file/functionName" -> "path/to/file:functionName"
        const filepath = functionNameWithSlashes.split("/").slice(0, -1).join("/");
        const functionName = functionNameWithSlashes.split("/").pop();
        return {
            udfPath: `${filepath}:${functionName}`,
            componentPath,
        };
    }
    throw new Error("Function address not supported");
}
async function getFunctionPathFromReference(functionReference) {
    // { name: "messages:list" }
    // { reference: "_reference/childComponent/aggregate/path/to/file/functionName" }
    // { functionHandle: "function://<id>/<path>" }
    const functionAddress = getFunctionAddress(functionReference);
    return await getFunctionPathFromAddress(functionAddress);
}
async function getFunctionFromPath(functionPath, type) {
    // "queries/messages:list" -> ["queries/messages", "list"]
    const [modulePath, maybeExportName] = functionPath.udfPath.split(":");
    const exportName = maybeExportName === undefined ? "default" : maybeExportName;
    const module = await getModulesForComponent(functionPath.componentPath)(modulePath);
    const func = module[exportName];
    if (func === undefined) {
        throw new Error(`Expected a Convex function exported from module "${modulePath}" as \`${exportName}\`, but there is no such export.`);
    }
    if (typeof func !== "function") {
        throw new Error(`Expected a Convex function exported from module "${modulePath}" as \`${exportName}\`, but got: ${func}`);
    }
    switch (type) {
        case "query":
            if (!func.isQuery) {
                throw new Error(`Expected a query function, but the function exported from module "${modulePath}" as \`${exportName}\` is not a query.`);
            }
            break;
        case "mutation":
            if (!func.isMutation) {
                throw new Error(`Expected a mutation function, but the function exported from module "${modulePath}" as \`${exportName}\` is not a mutation.`);
            }
            break;
        case "action":
            if (!func.isAction) {
                throw new Error(`Expected an action function, but the function exported from module "${modulePath}" as \`${exportName}\` is not an action.`);
            }
            break;
        case "any":
            break;
        default:
            throw type;
    }
    return func;
}
function simpleHash(string) {
    let hash = 0;
    for (let i = 0; i < string.length; i++) {
        const char = string.charCodeAt(i);
        hash = (hash << 5) - hash + char;
        hash = hash & hash; // Convert to 32bit integer
    }
    return hash;
}
